File: sdk/python/examples/transaction_sign.py
#!/usr/bin/env python

#
# Shows how to create all transactions manually using TransactionFactory.
#

import argparse
import importlib
from abc import abstractmethod
from binascii import hexlify, unhexlify

from symbolchain.CryptoTypes import PrivateKey
from symbolchain.facade.NemFacade import NemFacade
from symbolchain.facade.SymbolFacade import SymbolFacade


class TransactionSample:
	def __init__(self, facade):
		self.facade = facade
		self.key_pair = self.facade.KeyPair(PrivateKey(unhexlify('11002233445566778899AABBCCDDEEFF11002233445566778899AABBCCDDEEFF')))

	def process_transaction_descriptors(self, transaction_descriptors):
		for descriptor in transaction_descriptors:
			self.set_common_fields(descriptor)
			transaction = self.facade.transaction_factory.create(descriptor)
			self.sign_and_print(transaction)

	@abstractmethod
	def set_common_fields(self, descriptor):
		pass

	def sign_and_print(self, transaction):
		signature = self.facade.sign_transaction(self.key_pair, transaction)
		self.facade.transaction_factory.attach_signature(transaction, signature)

		print(f'Hash: {self.facade.hash_transaction(transaction)}')
		print(transaction)
		print(hexlify(transaction.serialize()))
		print('---- ' * 20)


class NemTransactionSample(TransactionSample):
	def __init__(self):
		super().__init__(NemFacade('testnet'))

	def set_common_fields(self, descriptor):
		descriptor.update({
			'signer_public_key': self.key_pair.public_key,
			'deadline': 12345
		})


class SymbolTransactionSample(TransactionSample):
	def __init__(self):
		super().__init__(SymbolFacade('testnet'))

	def set_common_fields(self, descriptor):
		descriptor.update({
			'signer_public_key': self.key_pair.public_key,
			'fee': 625,
			'deadline': 12345
		})


def main():
	parser = argparse.ArgumentParser(description='transaction sign example')
	parser.add_argument('--blockchain', help='blockchain', choices=('nem', 'symbol'), required=True)
	args = parser.parse_args()

	if 'nem' == args.blockchain:
		factory_names = [
			'nem_account_key_link',
			# 'nem_cosignature',
			'nem_mosaic',
			'nem_multisig_account',
			'nem_namespace',
			'nem_transfer'
		]
		sample = NemTransactionSample()
	else:
		factory_names = [
			'symbol_alias',
			'symbol_key_link',
			'symbol_lock',
			'symbol_metadata',
			'symbol_mosaic',
			'symbol_namespace',
			'symbol_restriction_account',
			'symbol_restriction_mosaic',
			'symbol_transfer'
		]
		sample = SymbolTransactionSample()

	total_descriptors_count = 0
	for factory_name in factory_names:
		transaction_descriptor_factory = getattr(importlib.import_module(f'examples.descriptors.{factory_name}'), 'descriptor_factory')
		transaction_descriptors = transaction_descriptor_factory()
		sample.process_transaction_descriptors(transaction_descriptors)
		total_descriptors_count += len(transaction_descriptors)

	print(f'finished processing {total_descriptors_count} descriptors')


if __name__ == '__main__':
	main()

File: sdk/javascript/examples/transaction_sign.js
#!/usr/bin/env node

//
// Shows how to create all transactions manually using TransactionFactory.
//

import symbolSdk from '../src/index.js';
import yargs from 'yargs';
import path from 'path';
import { fileURLToPath, pathToFileURL } from 'url';

(() => {
	class TransactionSample {
		constructor(facade, commonFields) {
			this.facade = facade;
			this.commonFields = commonFields;

			const privateKey = new symbolSdk.PrivateKey('11002233445566778899AABBCCDDEEFF11002233445566778899AABBCCDDEEFF');
			this.keyPair = new this.facade.constructor.KeyPair(privateKey);
		}

		processTransactionDescriptors(transactionDescriptors) {
			transactionDescriptors.forEach(descriptor => {
				this.setCommonFields(descriptor);
				const transaction = this.facade.transactionFactory.create(descriptor);
				this.signAndPrint(transaction);
			});
		}

		setCommonFields(descriptor) {
			Object.assign(descriptor, {
				signerPublicKey: this.keyPair.publicKey,
				...this.commonFields
			});
		}

		signAndPrint(transaction) {
			const signature = this.facade.signTransaction(this.keyPair, transaction);
			this.facade.transactionFactory.constructor.attachSignature(transaction, signature);

			console.log(`Hash: ${this.facade.hashTransaction(transaction)}`);
			console.log(transaction.toString());
			console.log('---- ---- ----');
		}
	}

	const runAllTests = (sample, factoryNames) => {
		let testsPending = 1;
		let totalDescriptorsCount = 0;

		const decrementTestsPending = () => {
			testsPending -= 1;
			if (0 === testsPending)
				console.log(`finished processing ${totalDescriptorsCount} descriptors`);
		};

		factoryNames.forEach(factoryName => {
			testsPending += 1;

			const filepath = path.join(path.dirname(fileURLToPath(import.meta.url)), 'descriptors', `${factoryName}.js`);
			import(pathToFileURL(filepath)).then(module => {
				const transactionDescriptors = module.default();
				sample.processTransactionDescriptors(transactionDescriptors);
				totalDescriptorsCount += transactionDescriptors.length;
			}).finally(() => {
				decrementTestsPending();
			});
		});

		decrementTestsPending();
	};

	const nemTransactionSample = new TransactionSample(new symbolSdk.facade.NemFacade('testnet'), {
		deadline: 12345
	});

	const symbolTransactionSample = new TransactionSample(new symbolSdk.facade.SymbolFacade('testnet'), {
		fee: 625n,
		deadline: 12345n
	});

	const args = yargs(process.argv.slice(2))
		.option('blockchain', {
			describe: 'blockchain to run examples against',
			choices: ['nem', 'symbol'],
			require: true
		})
		.argv;

	if ('nem' === args.blockchain) {
		runAllTests(nemTransactionSample, [
			'nem_account_key_link',
			'nem_mosaic',
			'nem_multisig_account',
			'nem_namespace',
			'nem_transfer'
		]);
	} else {
		runAllTests(symbolTransactionSample, [
			'symbol_alias',
			'symbol_key_link',
			'symbol_lock',
			'symbol_metadata',
			'symbol_mosaic',
			'symbol_namespace',
			'symbol_restriction_account',
			'symbol_restriction_mosaic',
			'symbol_transfer'
		]);
	}
})();
